### ISSUE TITLE
`Style/IfUnlessModifier` incorrectly flags LHS literal regular expression with named capture groups

### ISSUE THREAD

**Issue by cupakromer** [rubocop-hq/rubocop#6007]
## Expected behavior

Using the default Rubocop config the following should be valid:

```ruby
some_string = 'Any string value'
if /\b(?<ending>\w+)\z/ =~ some_string
  puts "Captured: #{ending}"
end
```

This uses the regular expression named capture using the literal left-hand side syntax. This is special syntax which creates local variables for the named captures. Per the [Ruby 2.5.1 `Regexp` docs](https://ruby-doc.org/core-2.5.1/Regexp.html#class-Regexp-label-Capturing):

> When named capture groups are used with a literal regexp on the left-hand side of an expression and the `=~` operator, the captured text is also assigned to local variables with corresponding names.
>
> ```ruby
> /\$(?<dollars>\d+)\.(?<cents>\d+)/ =~ "$3.67" #=> 0
> dollars #=> "3"
> ```

## Actual behavior

Rubocop flags the `if` conditional as being a violation of `Style/IfUnlessModifier`:

```
example.rb:2:1: C: Style/IfUnlessModifier: Favor modifier if usage when having a single-line body. Another good alternative is the usage of control flow &&/||.
if /\b(?<ending>\w+)\z/ =~ some_string
^^
```

The problem is it's invalid Ruby to use this type of regular expression syntax with a single-line modifier:

```ruby
some_string = 'Any string value'
puts "Captured: #{ending}" if /\b(?<ending>\w+)\z/ =~ some_string
```

```
# => example.rb:2:in `<main>': undefined local variable or method `ending' for main:Object (NameError)
```

## Steps to reproduce the problem

- Copy the **Expected Behavior** example into a file `example.rb`
- Using the default Rubocop run: `rubocop example.rb`

## RuboCop version

```
$ rubocop -V
0.57.2 (using Parser 2.5.1.0, running on ruby 2.5.1 x86_64-darwin16)
```

**Comment by mikegee** [comment]
I confirmed the report.

I also asked Ruby to check the syntax after autocorrecting:

```
ruby -cW2 test/test.rb
test/test.rb:2: warning: assigned but unused variable - ending
Syntax OK
```

Running it gives the `NameError` (as reported):

```
ruby test/test.rb
Traceback (most recent call last):
test/test.rb:2:in `<main>': undefined local variable or method `ending' for main:Object (NameError)
```
---

I was wrong about the ASTs being identical.  The initial code is parsed with the `ending` inside the `puts` argument as a local variable. After autocorrect, Ruby thinks it is a method call.

**Comment by mikegee** [comment]
Here is the simplest version of this kind of issue I can concoct:

```
% irb -W0
>> if a = 'a'; a; end
"a"
>> b if b = 'b'
Traceback (most recent call last):
        2: from /usr/local/bin/irb:11:in `<main>'
        1: from (irb):2
NameError (undefined local variable or method `b' for main:Object)
```
Doesn't this seem like a bug in Ruby for treating these differently?

**Comment by cupakromer** [comment]
I'm not sure if that is a Ruby :bug: or not. As far as I'm aware that's how Ruby has always treated one-line statements where a variable is defined in the conditional. I don't have a deep understanding of why things work that way. My superficial thought process was simply Ruby parses left-to-right and the variable isn't defined until after it is used, thus causing the issue.
